Create Notification from Idler

ranakan19 opened the pull request:
This PR adds the mechanism to create a Notification CR from member-operator after the notification builder has been moved to toolchain-common

fix wrt : https://issues.redhat.com/browse/CRT-1418
related PRs : 
https://github.com/codeready-toolchain/toolchain-common/pull/251, 
https://github.com/codeready-toolchain/host-operator/pull/629,
https://github.com/codeready-toolchain/api/pull/307,
https://github.com/codeready-toolchain/host-operator/pull/633
https://github.com/codeready-toolchain/toolchain-e2e/pull/560

xcoulon commented on controllers/idler/idler_controller.go:
```diff
@@ -2,8 +2,11 @@ package idler
 	"time"
 
+	"github.com/codeready-toolchain/toolchain-common/pkg/cluster"
+
```
no need for a blank line

ranakan19 replied:
done in [777b019](https://github.com/codeready-toolchain/member-operator/pull/357/commits/777b019b9dc2a2da36de0b0e2754b8012b10f7d8) and sorted imports.

xcoulon commented on controllers/idler/idler_controller.go:
```diff
@@ -25,6 +28,8 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/manager"
 	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
```
same here: no need for a blank line (let's keep the imports grouped)

ranakan19 replied:
done in [777b019](https://github.com/codeready-toolchain/member-operator/pull/357/commits/777b019b9dc2a2da36de0b0e2754b8012b10f7d8)

xcoulon commented on controllers/idler/idler_controller.go:
```diff
@@ -83,6 +89,9 @@ func (r *Reconciler) Reconcile(ctx context.Context, request ctrl.Request) (ctrl.
 		return reconcile.Result{}, r.wrapErrorWithStatusUpdate(logger, idler, r.setStatusFailed, err,
 			"failed to ensure idling '%s'", idler.Name)
 	}
+	//if err := r.createNotification(logger, idler); err != nil {
```
should be removed if it's not used

ranakan19 replied:
done in [777b019](https://github.com/codeready-toolchain/member-operator/pull/357/commits/777b019b9dc2a2da36de0b0e2754b8012b10f7d8)

xcoulon commented on controllers/idler/idler_controller.go:
```diff
@@ -141,6 +155,64 @@ func (r *Reconciler) ensureIdling(logger logr.Logger, idler *toolchainv1alpha1.I
 
+func (r *Reconciler) createNotification(logger logr.Logger, idler *toolchainv1alpha1.Idler) error {
+	//Get the HostClient
+	hostCluster, ok := r.GetHostCluster()
```
wouldn't it be simpler if `r.GetHostCluster()` returns `hostCluster, err`?
In other words, in line 47 above:
```
GetHostCluster      func() (*cluster.CachedToolchainCluster, error)
```

xcoulon replied:
yes, indeed

xcoulon commented on controllers/idler/idler_controller.go:
```diff
@@ -141,6 +155,64 @@ func (r *Reconciler) ensureIdling(logger logr.Logger, idler *toolchainv1alpha1.I
+	return nil
+}
+
+func (r *Reconciler) getUserEmailFromUserSignup(hostCluster *cluster.CachedToolchainCluster, idler *toolchainv1alpha1.Idler) []string {
```
```suggestion
func (r *Reconciler) getUserEmailsFromIdler(hostCluster *cluster.CachedToolchainCluster, idler *toolchainv1alpha1.Idler) []string {
```

ranakan19 replied:
changed it to getUserEmailFromMUR in [777b019](https://github.com/codeready-toolchain/member-operator/pull/357/commits/777b019b9dc2a2da36de0b0e2754b8012b10f7d8)

xcoulon commented on controllers/idler/idler_controller.go:
```diff
@@ -141,6 +155,64 @@ func (r *Reconciler) ensureIdling(logger logr.Logger, idler *toolchainv1alpha1.I
+func (r *Reconciler) getUserEmailFromUserSignup(hostCluster *cluster.CachedToolchainCluster, idler *toolchainv1alpha1.Idler) []string {
+	var emails []string
+	//get NSTemplateSet from idler
+	owner, found := idler.GetLabels()[toolchainv1alpha1.OwnerLabelKey]
+	if found {
```
```suggestion
	if owner, found := idler.GetLabels()[toolchainv1alpha1.OwnerLabelKey]; found {
```

xcoulon commented on controllers/idler/idler_controller.go:
```diff
@@ -141,6 +155,64 @@ func (r *Reconciler) ensureIdling(logger logr.Logger, idler *toolchainv1alpha1.I
+	if found {
+		nsTemplateSet := &toolchainv1alpha1.NSTemplateSet{}
+		r.Client.Get(context.TODO(), types.NamespacedName{Name: owner}, nsTemplateSet)
+		//get MUR from NSTemplateSetSpec
```
```suggestion
		// iterate on space roles from NSTemplateSet
```

xcoulon commented on controllers/idler/idler_controller.go:
```diff
@@ -141,6 +155,64 @@ func (r *Reconciler) ensureIdling(logger logr.Logger, idler *toolchainv1alpha1.I
+		nsTemplateSet := &toolchainv1alpha1.NSTemplateSet{}
+		r.Client.Get(context.TODO(), types.NamespacedName{Name: owner}, nsTemplateSet)
+		//get MUR from NSTemplateSetSpec
+		spaceRoles := nsTemplateSet.Spec.SpaceRoles
+		var murs []string
+		for _, spaceRole := range spaceRoles {
```
```suggestion
		var murs []string
		for _, spaceRole := range nsTemplateSet.Spec.SpaceRoles {
```

ranakan19 replied:
done in [777b019](https://github.com/codeready-toolchain/member-operator/pull/357/commits/777b019b9dc2a2da36de0b0e2754b8012b10f7d8)

xcoulon commented on controllers/idler/idler_controller.go:
```diff
@@ -141,6 +155,64 @@ func (r *Reconciler) ensureIdling(logger logr.Logger, idler *toolchainv1alpha1.I
+	}
+	// add sending notification here
+	//check the condition on Idler if notification already sent
+	_, found := condition.FindConditionByType(idler.Status.Conditions, toolchainv1alpha1.IdlerActivatedNotificationCreated)
+	if !found {
```
can be simplified in a single line statement with:
```suggestion
	if _, found := condition.FindConditionByType(idler.Status.Conditions, toolchainv1alpha1.IdlerActivatedNotificationCreated); !found {
```

ranakan19 replied:
done in [777b019](https://github.com/codeready-toolchain/member-operator/pull/357/commits/777b019b9dc2a2da36de0b0e2754b8012b10f7d8)

xcoulon commented on controllers/idler/idler_controller.go:
```diff
@@ -141,6 +155,64 @@ func (r *Reconciler) ensureIdling(logger logr.Logger, idler *toolchainv1alpha1.I
+	_, found := condition.FindConditionByType(idler.Status.Conditions, toolchainv1alpha1.IdlerActivatedNotificationCreated)
+	if !found {
+		userEmails := r.getUserEmailFromUserSignup(hostCluster, idler)
+		// Only create a notification if not created before
```
I believe that this comment should be moved a few lines above, otherwise we (or at least, I) may think that the "creat(ing) the notification if not created before" is related to each user, not to the idler.

ranakan19 replied:
done in [777b019](https://github.com/codeready-toolchain/member-operator/pull/357/commits/777b019b9dc2a2da36de0b0e2754b8012b10f7d8)

xcoulon commented on controllers/idler/idler_controller_test.go:
```diff
@@ -8,6 +8,8 @@ import (
 	"time"
 
+	"github.com/codeready-toolchain/toolchain-common/pkg/cluster"
+
```
same here, let's keep the imports grouped

ranakan19 replied:
done in [777b019](https://github.com/codeready-toolchain/member-operator/pull/357/commits/777b019b9dc2a2da36de0b0e2754b8012b10f7d8) and sorted imports

xcoulon commented on controllers/idler/idler_controller_test.go:
```diff
@@ -627,3 +638,7 @@ func prepareReconcileWithPodsRunningTooLong(t *testing.T, idler toolchainv1alpha
 	return reconciler, req, cl, allCl
 }
+
+func newGetHostClusterReady(fakeClient client.Client) cluster.GetHostClusterFunc {
```
`getHostCluster` is simpler to read than `newGetHostClusterReady`
```suggestion
func getHostCluster(fakeClient client.Client) cluster.GetHostClusterFunc {
```

xcoulon replied:
by the way, you could probably also test the case where the returned func will return an error when trying to the Host cluster

ranakan19 replied:
renamed in 7ccbf5d

MatousJobanek commented on controllers/idler/idler_controller_test.go:
```diff
@@ -338,16 +340,20 @@ func TestEnsureIdlingFailed(t *testing.T) {
-			assertCanNotGetObject(&appsv1.StatefulSet{}, "can't get statefulset")
-			assertCanNotGetObject(&openshiftappsv1.DeploymentConfig{}, "can't get deploymentconfig")
-			assertCanNotGetObject(&corev1.ReplicationController{}, "can't get replicationcontroller")
+			assertCanNotGetObject(&appsv1.Deployment{}, "can't get deployment", false)
+			assertCanNotGetObject(&appsv1.ReplicaSet{}, "can't get replicaset", false)
+			assertCanNotGetObject(&appsv1.DaemonSet{}, "can't get daemonset", true)
+			assertCanNotGetObject(&batchv1.Job{}, "can't get job", false)
+			assertCanNotGetObject(&appsv1.StatefulSet{}, "can't get statefulset", false)
+			assertCanNotGetObject(&openshiftappsv1.DeploymentConfig{}, "can't get deploymentconfig", false)
+			assertCanNotGetObject(&corev1.ReplicationController{}, "can't get replicationcontroller", false)
```
I'm sorry, this test doesn't seem to be the right place for verifying the Notification CR creation - this test is focused on failures and it's not really clear why only for the `DeploymentSet` the notification CR should be created

ranakan19 replied:
you're right, done in [777b019](https://github.com/codeready-toolchain/member-operator/pull/357/commits/777b019b9dc2a2da36de0b0e2754b8012b10f7d8)

MatousJobanek commented on controllers/idler/idler_controller_test.go:
```diff
@@ -208,7 +210,7 @@ func TestEnsureIdling(t *testing.T) {
 				memberoperatortest.AssertThatIdler(t, idler.Name, cl).
 					TracksPods(podsTooEarlyToKill.allPods).
-					HasConditions(memberoperatortest.Running())
+					HasConditions(memberoperatortest.Running(), memberoperatortest.IdlerNotificationCreated())
```
we should also make sure that the Notification CR actually exists in the host client when it should, and that it doesn't when it should not.

ranakan19 commented on controllers/idler/idler_controller.go:
```diff
@@ -141,6 +155,37 @@ func (r *Reconciler) ensureIdling(logger logr.Logger, idler *toolchainv1alpha1.I
+	if !ok {
+		err := fmt.Errorf("unable to get the host cluster")
+		logger.Error(err, "host Cluster not found")
+		return err
```
done in [777b019](https://github.com/codeready-toolchain/member-operator/pull/357/commits/777b019b9dc2a2da36de0b0e2754b8012b10f7d8)

ranakan19 commented on controllers/idler/idler_controller.go:
```diff
@@ -39,6 +44,7 @@ type Reconciler struct {
 	Client              client.Client
 	Scheme              *runtime.Scheme
 	AllNamespacesClient client.Client
+	GetHostCluster      func() (*cluster.CachedToolchainCluster, bool)
```
done in [777b019](https://github.com/codeready-toolchain/member-operator/pull/357/commits/777b019b9dc2a2da36de0b0e2754b8012b10f7d8)

ranakan19 commented on controllers/idler/idler_controller.go:
```diff
@@ -141,6 +155,64 @@ func (r *Reconciler) ensureIdling(logger logr.Logger, idler *toolchainv1alpha1.I
+			hostCluster.Client.Get(context.TODO(), types.NamespacedName{Name: mur, Namespace: hostCluster.OperatorNamespace}, getMUR)
+			emails = append(emails, getMUR.Annotations[toolchainv1alpha1.MasterUserRecordEmailAnnotationKey])
+		}
+	}
```
done in [777b019](https://github.com/codeready-toolchain/member-operator/pull/357/commits/777b019b9dc2a2da36de0b0e2754b8012b10f7d8)

ranakan19 commented on controllers/idler/idler_controller.go:
```diff
@@ -141,6 +155,64 @@ func (r *Reconciler) ensureIdling(logger logr.Logger, idler *toolchainv1alpha1.I
+	owner, found := idler.GetLabels()[toolchainv1alpha1.OwnerLabelKey]
+	if found {
+		nsTemplateSet := &toolchainv1alpha1.NSTemplateSet{}
+		r.Client.Get(context.TODO(), types.NamespacedName{Name: owner}, nsTemplateSet)
```
done in [777b019](https://github.com/codeready-toolchain/member-operator/pull/357/commits/777b019b9dc2a2da36de0b0e2754b8012b10f7d8)

ranakan19 commented on controllers/idler/idler_controller.go:
```diff
@@ -141,6 +155,64 @@ func (r *Reconciler) ensureIdling(logger logr.Logger, idler *toolchainv1alpha1.I
+		// get MUR from host and use user email from annotations
+		for _, mur := range murs {
+			getMUR := &toolchainv1alpha1.MasterUserRecord{}
+			hostCluster.Client.Get(context.TODO(), types.NamespacedName{Name: mur, Namespace: hostCluster.OperatorNamespace}, getMUR)
```
done in [777b019](https://github.com/codeready-toolchain/member-operator/pull/357/commits/777b019b9dc2a2da36de0b0e2754b8012b10f7d8)

ranakan19 commented:
/retest

rajivnathan commented on controllers/idler/idler_controller_test.go:
```diff
@@ -8,9 +8,15 @@ import (
+	"github.com/codeready-toolchain/toolchain-common/pkg/condition"
+
 	toolchainv1alpha1 "github.com/codeready-toolchain/api/api/v1alpha1"
+	nstemplatesetTest "github.com/codeready-toolchain/member-operator/controllers/nstemplateset"
```
Go package naming conventions discourage the use of underscores and mixedCaps. In this case it doesn't look like `nstemplateset` is colliding with other names in the file so can we just remove the alias?

rajivnathan commented on controllers/idler/idler_controller_test.go:
```diff
@@ -256,6 +262,116 @@ func TestEnsureIdling(t *testing.T) {
+		require.NoError(t, err)
+		assert.True(t, res.Requeue)
+		assert.Equal(t, int(res.RequeueAfter), 0)
+		// Idler tracks all pods now but pods have not been deleted yet
+		memberoperatortest.AssertThatInIdleableCluster(t, allCl).
+			PodsExist(podsRunningForTooLong.standalonePods).
+			DaemonSetExists(podsRunningForTooLong.daemonSet).
+			JobExists(podsRunningForTooLong.job).
+			DeploymentScaledUp(podsRunningForTooLong.deployment).
+			ReplicaSetScaledUp(podsRunningForTooLong.replicaSet).
+			DeploymentConfigScaledUp(podsRunningForTooLong.deploymentConfig).
+			ReplicationControllerScaledUp(podsRunningForTooLong.replicationController).
+			StatefulSetScaledUp(podsRunningForTooLong.statefulSet)
```
I think we can remove this since we're not really testing this behaviour in this testcase and it's covered by the other unit tests. Removing it would reduce noise and keep focus on what we really want to test.
```suggestion
```

rajivnathan commented on controllers/idler/idler_controller_test.go:
```diff
@@ -256,6 +262,116 @@ func TestEnsureIdling(t *testing.T) {
+		require.NoError(t, err)
+		assert.True(t, res.Requeue)
+		assert.Equal(t, int(res.RequeueAfter), 0)
+		// Idler tracks all pods now but pods have not been deleted yet
+		memberoperatortest.AssertThatInIdleableCluster(t, allCl).
+			PodsExist(podsRunningForTooLong.standalonePods).
+			DaemonSetExists(podsRunningForTooLong.daemonSet).
+			JobExists(podsRunningForTooLong.job).
+			DeploymentScaledUp(podsRunningForTooLong.deployment).
+			ReplicaSetScaledUp(podsRunningForTooLong.replicaSet).
+			DeploymentConfigScaledUp(podsRunningForTooLong.deploymentConfig).
+			ReplicationControllerScaledUp(podsRunningForTooLong.replicationController).
+			StatefulSetScaledUp(podsRunningForTooLong.statefulSet)
```
Same here
```suggestion
```

rajivnathan reviewed (commented):
Looks good overall, just left a few comments. Also, I think it would be good to cover 2 more cases in the unit tests:
1. When the Notification already exists when we're trying to create it
2. When the status update fails

ranakan19 commented:
/retest -updated e2e

ranakan19 commented:
/retest

ranakan19 commented on controllers/idler/idler_controller.go:
```diff
@@ -126,9 +135,14 @@ func (r *Reconciler) ensureIdling(logger logr.Logger, idler *toolchainv1alpha1.I
 					}
 					podLogger.Info("Pod deleted")
 				}
+				if _, err := r.createNotification(logger, idler); err != nil {
```
Don't know if its a good enough reason, but returning the boolean made it easier to test the function.

ranakan19 commented on controllers/idler/idler_controller.go:
```diff
@@ -141,6 +155,76 @@ func (r *Reconciler) ensureIdling(logger logr.Logger, idler *toolchainv1alpha1.I
+		}
+		if len(userEmails) > 0 {
+			for _, userEmail := range userEmails {
+				_, err := notify.NewNotificationBuilder(hostCluster.Client, hostCluster.OperatorNamespace).
```
To answer your question - if the notification was already but the status wasn't set, in the next reconcile it would create the notification again and try setting the status. This is because the status update is the only way currently that it is keeping track of whether a notification was sent, so a false condition would imply notification wasn't created even if it was.
I can improve message in condition such that the creation of notification looks at the message and reason of the condition as well, not just the status.
But, to confirm, isn't the situation when the Idler status is unable to update a rare case? Like Unable to Get a resource?

ranakan19 commented:
/retest

ranakan19 commented:
/retest

ranakan19 commented:
/retest

ranakan19 commented:
/retest

rajivnathan commented on controllers/idler/idler_controller.go:
```diff
@@ -141,6 +153,90 @@ func (r *Reconciler) ensureIdling(logger logr.Logger, idler *toolchainv1alpha1.I
+			getMUR := &toolchainv1alpha1.MasterUserRecord{}
+			err := hostCluster.Client.Get(context.TODO(), types.NamespacedName{Name: mur, Namespace: hostCluster.OperatorNamespace}, getMUR)
+			if err != nil {
+				return emails, fmt.Errorf("Could not get the MUR with name: %s ", mur)
```
By returning a new error here we're losing the error that came from the client so either we should just return the error from the client or wrap the error with our own error msg. Also, if we're wrapping the error with another message then same as previous comment it should start with a lowercase letter.

ranakan19 replied:
wrapped the error and changed to lowercase in [67a0dde](https://github.com/codeready-toolchain/member-operator/pull/357/commits/67a0dde897ead65c14667363a42ff694ae962c68)

rajivnathan commented on controllers/idler/idler_controller_test.go:
```diff
@@ -256,6 +266,95 @@ func TestEnsureIdling(t *testing.T) {
+			Namespace: test.HostOperatorNs,
+			Name:      "alex-stage-idled",
+		}, notification)
+		require.NoError(t, err)
```
We should also verify the notification spec and labels values are as expected

ranakan19 replied:
done in a58c63e

rajivnathan commented on controllers/idler/idler_controller_test.go:
```diff
@@ -256,6 +266,95 @@ func TestEnsureIdling(t *testing.T) {
+			Name:      "alex-stage-idled",
+		}, notification)
+		require.NoError(t, err)
+
```
We should also add another reconcile to ensure the notification is not recreated/changed

ranakan19 replied:
I do check in a separate test case that the notification is not recreated - https://github.com/ranakan19/member-operator/blob/a58c63eb12b9f083fe3273d956221f77d8980f9e/controllers/idler/idler_controller_test.go#L541

rajivnathan commented on controllers/idler/idler_controller_test.go:
```diff
@@ -256,6 +266,95 @@ func TestEnsureIdling(t *testing.T) {
+		require.Contains(t, err.Error(), "\"alex-dev-idled\" not found")
+
+		t.Run("third reconcile doesn't create notification as no pods to idle", func(t *testing.T) {
+			// second reconcile should delete pods and create notification
```
Looks like this comment needs to be updated if we end up needing the 3rd reconcile

ranakan19 replied:
removed the entire test case, done in a58c63e

rajivnathan commented on controllers/idler/idler_controller_test.go:
```diff
@@ -256,6 +266,95 @@ func TestEnsureIdling(t *testing.T) {
+		require.Error(t, err)
+		require.Contains(t, err.Error(), "\"alex-dev-idled\" not found")
+
+		t.Run("third reconcile doesn't create notification as no pods to idle", func(t *testing.T) {
```
Do we really need another reconcile? Is this step just supposed to verify the idler status is set correctly? If so wouldn't the status be set after the 2nd reconcile so we could just add that check right after the 2nd reconcile?

ranakan19 replied:
This entire test case has been removed, it did not really add anything to the coverage. done in a58c63e

rajivnathan commented on controllers/idler/idler_controller_test.go:
```diff
@@ -448,6 +547,188 @@ func TestEnsureIdlingFailed(t *testing.T) {
+		require.True(t, condition.IsTrue(idler.Status.Conditions, toolchainv1alpha1.IdlerTriggeredNotificationCreated))
+	})
+
+	t.Run("Condition is set when setting failed previously, and notification is only sent once", func(t *testing.T) {
```
Thanks for covering this case 👍

rajivnathan commented on controllers/idler/idler_controller_test.go:
```diff
@@ -448,6 +547,188 @@ func TestEnsureIdlingFailed(t *testing.T) {
+		err := reconciler.createNotification(logf.FromContext(context.TODO()), idler)
+
+		//then
+		require.Error(t, err, "can't update condition")
```
```suggestion
		require.EqualError(t, err, "can't update condition")
```

ranakan19 replied:
done in [67a0dde](https://github.com/codeready-toolchain/member-operator/pull/357/commits/67a0dde897ead65c14667363a42ff694ae962c68)

rajivnathan commented on controllers/idler/idler_controller.go:
```diff
@@ -141,6 +153,90 @@ func (r *Reconciler) ensureIdling(logger logr.Logger, idler *toolchainv1alpha1.I
+		nsTemplateSet := &toolchainv1alpha1.NSTemplateSet{}
+		err := r.Client.Get(context.TODO(), types.NamespacedName{Name: owner, Namespace: r.Namespace}, nsTemplateSet)
+		if err != nil {
+			logger.Error(err, "Could not get the NSTemplateSet with name", "owner", owner)
```
Error messages should start with a lowercase according to golang conventions

ranakan19 replied:
corrected in [67a0dde](https://github.com/codeready-toolchain/member-operator/pull/357/commits/67a0dde897ead65c14667363a42ff694ae962c68)

rajivnathan commented on controllers/idler/idler_controller_test.go:
```diff
@@ -448,6 +547,188 @@ func TestEnsureIdlingFailed(t *testing.T) {
+		//when
+		err := reconciler.createNotification(logf.FromContext(context.TODO()), idler)
+		//then
+		require.Error(t, err)
```
Let's also verify the expected error message

rajivnathan reviewed (commented):
Nice work! 👍  Left some comments/suggestions but looks good overall. Please also check the test coverage for missing unit tests for things like error scenarios

xcoulon commented on controllers/idler/idler_controller.go:
```diff
@@ -141,6 +153,90 @@ func (r *Reconciler) ensureIdling(logger logr.Logger, idler *toolchainv1alpha1.I
+			if err != nil {
+				return emails, fmt.Errorf("Could not get the MUR with name: %s ", mur)
+			}
+			emails = append(emails, getMUR.Annotations[toolchainv1alpha1.MasterUserRecordEmailAnnotationKey])
```
it should not happen, but just in case, you could check that the `toolchainv1alpha1.MasterUserRecordEmailAnnotationKey` annotation is not empty

ranakan19 replied:
done in a58c63e

xcoulon commented on controllers/idler/idler_controller.go:
```diff
@@ -141,6 +153,90 @@ func (r *Reconciler) ensureIdling(logger logr.Logger, idler *toolchainv1alpha1.I
+			emails = append(emails, getMUR.Annotations[toolchainv1alpha1.MasterUserRecordEmailAnnotationKey])
+		}
+	} else {
+		logger.Info(fmt.Sprintf("Idler %s does not have any owner label", idler.Name), "Owner Missing", idler)
```
I believe it's better to use "constant" error messages and use a separate field for the idler name, so we can more easily search in the logs
```suggestion
		logger.Info("Idler does not have any owner label", "idler_name", idler.Name)
```

ranakan19 commented on controllers/idler/idler_controller.go:
```diff
@@ -109,7 +115,7 @@ func (r *Reconciler) ensureIdling(logger logr.Logger, idler *toolchainv1alpha1.I
 	newStatusPods := make([]toolchainv1alpha1.Pod, 0, 10)
 	for _, pod := range podList.Items {
-		podLogger := logger.WithValues("pod_name", pod.Name, "pod_phase", pod.Status.Phase)
+		podLogger := logger.WithValues("fne", pod.Name, "pod_phase", pod.Status.Phase)
```
yes, thanks for catching this - done in [67a0dde](https://github.com/codeready-toolchain/member-operator/pull/357/commits/67a0dde897ead65c14667363a42ff694ae962c68)

ranakan19 commented on controllers/idler/idler_controller.go:
```diff
@@ -141,6 +153,90 @@ func (r *Reconciler) ensureIdling(logger logr.Logger, idler *toolchainv1alpha1.I
+		notification := &toolchainv1alpha1.Notification{}
+		// Check if notification already exists in host
+		if err := hostCluster.Client.Get(context.TODO(), types.NamespacedName{Name: notificationName, Namespace: hostCluster.OperatorNamespace}, notification); err != nil {
+			if errors.IsNotFound(err) {
```
Thank You @rajivnathan and @xcoulon ! It does simplify the function greatly

ranakan19 commented on controllers/idler/idler_controller.go:
```diff
@@ -141,6 +153,90 @@ func (r *Reconciler) ensureIdling(logger logr.Logger, idler *toolchainv1alpha1.I
+				if err != nil {
+					return err
+				}
+				if len(userEmails) > 0 {
```
done in [67a0dde](https://github.com/codeready-toolchain/member-operator/pull/357/commits/67a0dde897ead65c14667363a42ff694ae962c68)

ranakan19 commented on controllers/idler/idler_controller.go:
```diff
@@ -141,6 +153,90 @@ func (r *Reconciler) ensureIdling(logger logr.Logger, idler *toolchainv1alpha1.I
+}
+
+func (r *Reconciler) getUserEmailsFromMURs(logger logr.Logger, hostCluster *cluster.CachedToolchainCluster, idler *toolchainv1alpha1.Idler) ([]string, error) {
+	logger.Info("getUserEmailsFromMURs")
```
yep, removed in [67a0dde](https://github.com/codeready-toolchain/member-operator/pull/357/commits/67a0dde897ead65c14667363a42ff694ae962c68)

ranakan19 commented:
/retest

ranakan19 commented:
/retest

ranakan19 commented on controllers/idler/idler_controller.go:
```diff
@@ -141,6 +155,91 @@ func (r *Reconciler) ensureIdling(logger logr.Logger, idler *toolchainv1alpha1.I
+			if err != nil {
+				return emails, errs.Wrapf(err, "could not get the MUR")
+			}
+			if email := getMUR.Annotations[toolchainv1alpha1.MasterUserRecordEmailAnnotationKey]; email != "" {
```
it returns an error at line 186

rajivnathan replied:
The error at line 186 would happen only if there are no emails returned from this function. eg. the "Idler does not have any owner label" case. If there are multiple MURs and 1 of them doesn't have an email then we would send the email to  all except that one and there would be nothing wrong in the logs (we effectively ignored that one user). We should probably just log it as an error if the MUR is found but we cannot find an email for it but yeah, no need to return the error in this case I think.

rajivnathan reviewed (commented):
Thanks for addressing the comments! 🙌 

I think we're just missing test coverage for the following cases:
1. status update error
2. host client create error when creating notification

ranakan19 commented:
> ` [1c43fd0](/codeready-toolchain/member-operator/pull/357/commits/1c43fd095e26618b6b43ece751f36e21988e78ed)`

done in 1c43fd0

ranakan19 commented:
> Thanks for addressing the comments! 🙌
> 
> I think we're just missing test coverage for the following cases:
> 
> 1. status update error
> 2. host client create error when creating notification

@rajivnathan Thank you! done in 1c43fd0

rajivnathan commented on controllers/idler/idler_controller_test.go:
```diff
@@ -438,14 +506,256 @@ func TestEnsureIdlingFailed(t *testing.T) {
+
+		// then
+		// since no mur, error should have been wrapped as status update, but status update fails
+		require.EqualError(t, err, "failed to ensure idling 'john-dev': cannot set status to fail")
```
This just made me realize that we are returning an error if the status update fails and that would also stop the pod tracking so we should just log the error in this case too.

ranakan19 replied:
done in [f347b00](https://github.com/codeready-toolchain/member-operator/pull/357/commits/f347b0044fdde5e0f21331cdc292c5ba5ff8b409)

rajivnathan reviewed (approved):
Nicely done! 👍

alexeykazakov commented on controllers/idler/idler_controller.go:
```diff
@@ -141,6 +155,91 @@ func (r *Reconciler) ensureIdling(logger logr.Logger, idler *toolchainv1alpha1.I
+		return fmt.Errorf("unable to get the host cluster")
+	}
+	//check the condition on Idler if notification already sent, only create a notification if not created before
+	_, found := condition.FindConditionByType(idler.Status.Conditions, toolchainv1alpha1.IdlerTriggeredNotificationCreated)
+	if found && condition.IsTrue(idler.Status.Conditions, toolchainv1alpha1.IdlerTriggeredNotificationCreated) {
```
As far as I remember `IsTrue()` returns false if the condition is not found.
```suggestion
	if condition.IsTrue(idler.Status.Conditions, toolchainv1alpha1.IdlerTriggeredNotificationCreated) {
```
